kubespawner/clients.py

```diff
@@ -0,0 +1,27 @@
+import weakref
+
+import kubernetes.client
+
+_client_cache = {}
+
+
+def shared_client(ClientType, *args, **kwargs):
+    """Return a single shared kubernetes client instance
+
+    A weak reference to the instance is cached,
+    so that concurrent calls to shared_client
+    will all return the same instance until
+    all references to the client are cleared.
+    """
+    kwarg_key = tuple((key, kwargs[key]) for key in sorted(kwargs))
+    cache_key = (ClientType, args, kwarg_key)
+    client = None
+    if cache_key in _client_cache:
+        client = _client_cache[cache_key]()
```

Comment:
Why the trailing `()`? I think `_client_cache` contains instances already so this would end up calling `__call__` on that instance.

Reply:
`_client_cache` contains weak references to instances, not the instances themselves. This allows them to be garbage collected. The ref must be called in order to return the instance. If it has been garbage collected in the meantime, resolving the weakref will return None.

I've added a comment to clarify what's happening here.

I've